### ISSUE TITLE
Plugin for adding a custom keyboard shortcut

### ISSUE THREAD

**Issue by rcywongaa** [gnome-terminator/terminator#629]
**Is your feature request related to a problem? Please describe.**
This is a follow up for [this question on StackOverflow](https://stackoverflow.com/questions/73271332/how-to-create-a-plugin-that-adds-a-custom-keyboard-shortcut).  In order to create a plugin that adds a custom keyboard shortcut, there needs to be a way to detect when a new terminator window is created and trigger `window.connect('key-press-event', self.onKeyPress)`.  From my limited knowledge, I don't think this is possible with the current plugin system

**Describe the solution you'd like**
A plugin type for creating custom keyboard shortcut, or more generally a plugin type that can detect new window events

**Additional context**
I am trying to create a plugin that upon detecting a certain keypress, copies the current scrollback buffer and opens it in vim.  Currently, I am able to trigger this on the first terminator window, but subsequent terminator windows don't detect the key press.

**Comment by mattrose** [comment]
I think I see what your problem is.  I'll take a look at this when I get the chance.  It may be just drilling down in the code to get the correct terminal to pass the keypress action to, or it may mean editing the plugin code.

**Comment by vssdeo** [comment]
Not sure this might help but was playing with this and seems for me if I comment the dbus exception part which stops the processing of config/plugin etc. This works for new windows. I would still need to dig deep but this info may help, who knows. 

```
diff --git a/terminatorlib/ipc.py b/terminatorlib/ipc.py
index 4303f28f..32201b72 100644
--- a/terminatorlib/ipc.py
+++ b/terminatorlib/ipc.py
@@ -62,8 +62,8 @@ class DBusService(Borg, dbus.service.Object):
             flags = 1 | 4 # allow replacement | do not queue
             if not proxy.RequestName(BUS_NAME, dbus.UInt32(flags)) in (1, 4):
                 dbg('bus name unavailable: %s' % BUS_NAME)
-                raise dbus.exceptions.DBusException(
-                    "Couldn't get DBus name %s: Name exists" % BUS_NAME)
+                #raise dbus.exceptions.DBusException(
+                #    "Couldn't get DBus name %s: Name exists" % BUS_NAME)

```

**Comment by vssdeo** [comment]
Hi.

Tried to understand the dbus issue, I hope I am on right track. If so can push the change. It seems each new window session tries to find the name of any dbus and tries to get its name.

First refering Dbus Ref: https://dbus.freedesktop.org/doc/api/html/dbus-shared_8h_source.html

/* Owner flags */
#define DBUS_NAME_FLAG_ALLOW_REPLACEMENT 0x1 
#define DBUS_NAME_FLAG_REPLACE_EXISTING  0x2 
#define DBUS_NAME_FLAG_DO_NOT_QUEUE      0x4 

/* Replies to request for a name */
#define DBUS_REQUEST_NAME_REPLY_PRIMARY_OWNER  1 
#define DBUS_REQUEST_NAME_REPLY_IN_QUEUE       2 
#define DBUS_REQUEST_NAME_REPLY_EXISTS         3 
#define DBUS_REQUEST_NAME_REPLY_ALREADY_OWNER  4 


It seems in our terminator code the line :
     `proxy.RequestName(BUS_NAME, dbus.UInt32(flags)))` is returning DBUS_REQUEST_NAME_REPLY_EXISTS  value (3) 

And we are checking it against return values (1 ,4) (DBUS_REQUEST_NAME_REPLY_PRIMARY_OWNER, DBUS_REQUEST_NAME_REPLY_ALREADY_OWNER) in the line.

     if not proxy.RequestName(BUS_NAME, dbus.UInt32(flags)) in (1, 4):

where as its returning value (3) DBUS_REQUEST_NAME_REPLY_EXISTS which seems to be right as there is another window instance running, so shouldn't we be reusing the same name and checking like (1, 3, 4)

                   if not proxy.RequestName(BUS_NAME, dbus.UInt32(flags)) in (1, 3, 4):

and in logs then I can see the dbus name take on the line:
 ```
self.bus_name = dbus.service.BusName(BUS_NAME, bus=dbus.SessionBus())
```

DBusService::prepare_attributes: MYDEBUG <dbus.service.BusName net.tenshu.Terminator21a9d5db22c73a993ff0b42f64b396873 on <dbus._dbus.SessionBus (session) at 0x7f0c055fd990> at 0x7f0c059f34c0> (/home/user/dev/terminator/terminatorlib/ipc.py:70)

**Comment by vssdeo** [comment]
Yes that's the part which is making me wonder for its not the plugin, but it happens every time when a new window is opened. Like you said it should instead use the same bus. So on my master without my change (above) and without plugin, when I launch two separate windows from command line. On the second one I have been seeing.
```
----logs---
DBusService::prepare_attributes: Checking for bus name availability: net.tenshu.Terminator21a9d5db22c73a993ff0b42f64b396873 (/home/user/dev/terminator/terminatorlib/ipc.py:54)
DBusService::prepare_attributes: bus name unavailable: net.tenshu.Terminator21a9d5db22c73a993ff0b42f64b396873 (/home/user/dev/terminator/terminatorlib/ipc.py:78)
noclass::<module>: Unable to become master process, operating via DBus (/home/user/dev/terminator/./terminator:115)
noclass::<module>: Requesting a new window (/home/user/dev/terminator/./terminator:130)
----
```

and so the bus already exists then shouldn't it REUSE the same bus again here are the next lines in ipy.py, original code without adding DBUS_REQUEST_NAME_REPLY_EXISTS (3)

 
 ```
 if not proxy.RequestName(BUS_NAME, dbus.UInt32(flags)) in (1, 4):
                dbg('bus name unavailable: %s' % BUS_NAME)
                raise dbus.exceptions.DBusException(
                    "Couldn't get DBus name %s: Name exists" % BUS_NAME)
            self.bus_name = dbus.service.BusName(BUS_NAME,
                                                 bus=dbus.SessionBus())

```

This has been making me wonder since I first started to read the code (if I am not wrong) and there are other aspect of ipc.py which I am trying to understand but questions on it may be later.

**Comment by vssdeo** [comment]
It seems you are right, my change does break a behavior. When I launch  two instances ./terminator and  second ./terminator --new-tab the second one does not seem to attach to first. Let me do more digging.

**Comment by vssdeo** [comment]
Looking at this fresh, in terms of new window creation, seems that the signal connect which happens in the plugin __init__ is not copied to the new window and since the plugin.py has already loaded the plugins and called __init__ previously, It can't connect again.

The config reload in plugin.py load_plugins(self) stops at check:
```
 if self.done:
         dbg('Already loaded')
         return
```
and further in same function at: since init is already done of plugins.
```
 if item not in self.instances:
         self.instances[item] = func()
```

re-init seems to work but with loading plugins again. Will check if signals can be copied  from one window to another.

**Comment by vssdeo** [comment]
Apart from reloading & re-init plugins, I could only think of copying signals, and was not able to find a gtk supported way, but attached is a patch which is a demo. This overrides the connect call for window and copies other signals (non window preset signals) from last created window, not in new window. Will keep thinking on the best way out. Attachment:
[signal_patch.txt](https://github.com/gnome-terminator/terminator/files/9927848/signal_patch.txt)

**Comment by mattrose** [comment]
@vssdeo Why not reload and re-initialize plugins when a new window is created?  This is the most natural solution, IMO.

**Comment by vssdeo** [comment]
@mattrose yup but I still wanted to put all options on table :) and take an opinion. For new instance seems natural.

**Comment by vssdeo** [comment]
As per the discussion https://github.com/gnome-terminator/terminator/pull/682 and since I had given a patch for signal copying signal_patch.txt (above) I am attaching the plugin reload test patch (in case others want to experiment), this will put both these solutions in a single thread and for record. 

**Note:** This is a basic test patch not final changes.

[plugin_reload_test_diff_patch.txt](https://github.com/gnome-terminator/terminator/files/10162932/plugin_reload_test_diff_patch.txt)

**Comment by vssdeo** [comment]
custom keybindings have been added : https://github.com/gnome-terminator/terminator/issues/681 Can be closed.
